src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/SnowplowEvent.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2016-2019 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.analytics.scalasdk
+
+// circe
+import io.circe.syntax._
+import io.circe.{Encoder, Json, JsonObject}
+
+// iglu
+import com.snowplowanalytics.iglu.core.circe.CirceIgluCodecs._
+import com.snowplowanalytics.iglu.core.{SchemaKey, SelfDescribingData}
+
+object SnowplowEvent {
+
+  /**
+    * A JSON representation of an atomic event's unstruct_event field.
+    *
+    * @param data the unstruct event as self-describing JSON, or None if the field is missing
+    */
+  case class UnstructEvent(data: Option[SelfDescribingData[Json]]) extends AnyVal {
+    def toJson: Option[(String, Json)] = {
+      data.map {
+        case SelfDescribingData(s, d) =>
+          (transformSchema(Data.UnstructEvent, s.vendor, s.name, s.version.model), d)
+      }
+    }
+  }
+
+  /**
+    * A JSON representation of an atomic event's contexts or derived_contexts fields.
+    *
+    * @param data the context as self-describing JSON, or None if the field is missing
+    */
+  case class Contexts(data: List[SelfDescribingData[Json]]) extends AnyVal {
+    def toJson: Map[String, Json] = {
+      data.groupBy(x => (x.schema.vendor, x.schema.name, x.schema.format, x.schema.version.model)).map {
+        case ((vendor, name, _, model), d) =>
+          (transformSchema(Data.Contexts(Data.CustomContexts), vendor, name, model), d.map {
+            selfdesc => selfdesc.data
+          }.asJson)
+      }
+    }
+  }
+
+  implicit final val contextsCirceEncoder: io.circe.Encoder[Contexts] =
+    Encoder.instance { contexts: Contexts =>
+      if (contexts.data.isEmpty) List[String]().asJson
```

Comment:
Still wondering why for empty contexts we return an *array*, but for non-empty an *object*?

Also, `SelfDescribingData(Common.ContextsUri, contexts.data).asJson` (or `.normalize`) looks a little bit more readable to me.

Reply:
The second suggestion would require an encoder for `SelfDescribingData[List[SelfDescribingData[Json]]]`.

Reply:
`List[SelfDescribingData[Json]]` is isomorphic to `Json`, because it will be just an array of self-describing JSONs. I get your point that we don't win in readability much, but still I see it as a win.